File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/forge/ForgeInstallTask.java
package org.wdt.wdtc.core.download.forge;


import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.apache.log4j.Logger;
import org.wdt.utils.dependency.DefaultDependency;
import org.wdt.utils.dependency.DependencyDownload;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.utils.io.FilenameUtils;
import org.wdt.wdtc.core.download.SpeedOfProgress;
import org.wdt.wdtc.core.download.game.DownloadGameClass;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.GameVersionJsonObject;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.LibraryObject;
import org.wdt.wdtc.core.game.config.DefaultGameConfig;
import org.wdt.wdtc.core.manger.DownloadSourceManger;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.manger.URLManger;
import org.wdt.wdtc.core.utils.*;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.jar.JarInputStream;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ForgeInstallTask extends ForgeDownloadInfo implements InstallTaskInterface {
  private static final Logger logmaker = WdtcLogger.getLogger(ForgeInstallTask.class);
  private final DefaultGameConfig.Config config;

  public ForgeInstallTask(Launcher launcher, String forgeVersion) {
    super(launcher, forgeVersion);
    this.config = launcher.getGameConfig().getConfig();
  }

  public ForgeInstallTask(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    this(launcher, versionJsonObjectInterface.getVersionNumber());
  }

  public String CommandLine(int index) throws IOException {
    JsonObject JsonObject = getInstallPrefileJSONObject().getAsJsonArray("processors").get(index).getAsJsonObject();
    StringBuilder CommandLine = new StringBuilder();
    CommandLine.append(config.getJavaPath()).append(" -cp ");
    JsonArray JarList = JsonObject.getAsJsonArray("classpath");
    for (int i = 0; i < JarList.size(); i++) {
      DependencyDownload Jar = new DependencyDownload(JarList.get(i).getAsString());
      Jar.setDownloadPath(launcher.getGameLibraryDirectory());
      CommandLine.append(FilenameUtils.separatorsToWindows(Jar.getLibraryFilePath())).append(";");
    }
    DependencyDownload MainJar = new DependencyDownload(JsonObject.get("jar").getAsString());
    MainJar.setDownloadPath(launcher.getGameLibraryDirectory());
    String MainClass = new JarInputStream(new FileInputStream(MainJar.getLibraryFile())).getManifest().getMainAttributes().getValue("Main-Class");
    CommandLine.append(MainJar.getLibraryFilePath()).append(" ").append(MainClass).append(" ");
    JsonArray ArgsList = JsonObject.getAsJsonArray("args");
    for (int i = 0; i < ArgsList.size(); i++) {
      if (i % 2 == 0) {
        CommandLine.append(ArgsList.get(i).getAsString()).append(" ");
      } else {
        String ArgeStr = ArgsList.get(i).getAsString();
        Matcher Middle = getMiddleBracket(ArgeStr);
        Matcher Large = getLargeBracket(ArgeStr);
        if (ArgeStr.equals("{MINECRAFT_JAR}")) {
          CommandLine.append(launcher.getVersionJar()).append(" ");
        } else if (ArgeStr.equals("{BINPATCH}")) {
          ZipUtils.unZipBySpecifyFile(getForgeInstallJarPath(), getClientLzmaFile());
          CommandLine.append(getClientLzmaFile()).append(" ");
        } else if (ArgeStr.equals("{SIDE}")) {
          CommandLine.append("client").append(" ");
        } else if (Large.find()) {
          DependencyDownload client = new DependencyDownload(Clean(getInstallPrefileJSONObject().getAsJsonObject("data")
              .getAsJsonObject(Large.group(1)).get("client").getAsString()));
          client.setDownloadPath(launcher.getGameLibraryDirectory());
          CommandLine.append(client.getLibraryFilePath()).append(" ");
        } else if (Middle.find()) {
          DependencyDownload arge = new DependencyDownload(Middle.group(1));
          arge.setDownloadPath(launcher.getGameLibraryDirectory());
          CommandLine.append(arge.getLibraryFilePath()).append(" ");
        } else {
          CommandLine.append(ArgeStr).append(" ");
        }
      }
    }
    return FilenameUtils.separatorsToWindows(CommandLine.toString());
  }

  public void DownloadClientText() throws IOException {
    DefaultDependency TxtPath = null;
    JsonObject DataObject = getInstallPrefileJSONObject().getAsJsonObject("data");
    if (DataObject.has("MOJMAPS")) {
      Matcher matcher = getMiddleBracket(DataObject.getAsJsonObject("MOJMAPS").get("client").getAsString());
      if (matcher.find()) {
        TxtPath = new DefaultDependency(matcher.group(1));
      }
    } else {
      Matcher matcher = getMiddleBracket(DataObject.getAsJsonObject("MAPPINGS").get("client").getAsString());
      if (matcher.find()) {
        TxtPath = new DefaultDependency(matcher.group(1));
      }
    }
    String TxtUrl = JsonUtils.getJsonObject(launcher.getVersionJson()).getAsJsonObject("downloads")
        .getAsJsonObject("client_mappings").get("url").getAsString();
    if (DownloadSourceManger.isNotOfficialDownloadSource()) {
      TxtUrl = URLUtils.getRedirectUrl(TxtUrl.replaceAll(URLManger.getPistonDataMojang(), source.getDataUrl()));
    }
    if (TxtPath != null) {
      DownloadUtils.StartDownloadTask(TxtUrl, launcher.getGameLibraryDirectory() + TxtPath.formJar());
    }

  }

  public void InstallForge() throws IOException {
    JsonArray objects = getInstallPrefileJSONObject().getAsJsonArray("processors");
    for (int i = 0; i < objects.size(); i++) {
      JsonObject TaskJson = objects.get(i).getAsJsonObject();
      if (TaskJson.has("sides")) {
        if (TaskJson.getAsJsonArray("sides").get(0).getAsString().equals("client")) {
          StartCommand(i);
        }
      } else {
        if (!TaskJson.getAsJsonArray("args").get(1).getAsString().equals("DOWNLOAD_MOJMAPS")) {
          StartCommand(i);
        }
      }
    }
  }

  public File getClientLzmaFile() {
    return new File(FileManger.getWdtcCache(), "/data/client.lzma");
  }

  public String Clean(String str) {
    return StringUtils.cleanStrInString(str, "[", "]", "{", "}");
  }

  private void StartCommand(int i) throws IOException {
    String commmand = CommandLine(i);
    logmaker.info("Command Line:" + commmand);
    Process process = Runtime.getRuntime().exec(new String[]{"cmd", "/c", commmand});
    BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(process.getInputStream(), "GBK"));
    String line;
    while ((line = bufferedReader.readLine()) != null) {
      logmaker.info(line);
    }
    process.destroy();
  }

  public Matcher getMiddleBracket(String args) {
    return Pattern.compile("\\[(.+)]").matcher(args);
  }

  public Matcher getLargeBracket(String args) {
    return Pattern.compile("\\{(.+)}").matcher(args);
  }

  @Override
  public void overwriteVersionJson() throws IOException {
    JsonObject ForgeVersionJsonObject = getForgeVersionJsonObject();
    JsonArray LibraryArray = ForgeVersionJsonObject.getAsJsonArray("libraries");
    GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
    GameVersionJsonObject.Arguments arguments = VersionJsonObject.getArguments();
    JsonArray GameList = arguments.getGameList();
    JsonObject Arguments = ForgeVersionJsonObject.getAsJsonObject("arguments");
    GameList.addAll(Arguments.getAsJsonArray("game"));
    arguments.setGameList(GameList);
    JsonArray JvmList = arguments.getJvmList();
    if (Arguments.has("jvm")) {
      JvmList.addAll(Arguments.getAsJsonArray("jvm"));
    }
    arguments.setJvmList(JvmList);
    VersionJsonObject.setArguments(arguments);
    for (int i = 0; i < LibraryArray.size(); i++) {
      LibraryObject libraryObject = JsonObjectUtils.parseObject(LibraryArray.get(i).getAsString(), LibraryObject.class);
      VersionJsonObject.getLibraries().add(libraryObject);
    }
    VersionJsonObject.setMainClass(ForgeVersionJsonObject.get("mainClass").getAsString());
    VersionJsonObject.setId(launcher.getVersionNumber() + "-forge-" + ForgeVersionNumber);
    launcher.putToVersionJson(VersionJsonObject);
  }

  @Override
  public void writeVersionJsonPatches() throws IOException {
    GameVersionJsonObject Object = launcher.getGameVersionJsonObject();
    List<JsonObject> ObjectList = new ArrayList<>();
    ObjectList.add(JsonUtils.getJsonObject(launcher.getVersionJson()));
    ObjectList.add(JsonUtils.getJsonObject(getForgeVersionJsonPath()));
    Object.setJsonObject(ObjectList);
    launcher.putToVersionJson(Object);
  }

  @Override
  public void afterDownloadTask() throws IOException {
    DownloadForgeLibraryFile(getInstallProfilePath());
    DownloadClientText();
    InstallForge();
  }

  @Override
  public void beforInstallTask() {
    DownloadInstallJar();
    getInstallProfile();
    getForgeVersionJson();
  }

  public void DownloadForgeLibraryFile(File file) throws IOException {
    JsonArray LibraryList = JsonUtils.getJsonObject(file).getAsJsonArray("libraries");
    SpeedOfProgress speed = new SpeedOfProgress(LibraryList.size());
    for (int i = 0; i < LibraryList.size(); i++) {
      LibraryObject object = LibraryObject.getLibraryObject(LibraryList.get(i).getAsJsonObject());
      DownloadGameClass task = new DownloadGameClass(launcher);
      task.StartDownloadLibraryTask(object, speed);
    }
    speed.await();
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/forge/ForgeInstallTask.kt
package org.wdt.wdtc.core.download.forge

import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import org.wdt.utils.gson.getJsonObject
import org.wdt.utils.gson.parseObject
import org.wdt.utils.gson.readFileToJsonObject
import org.wdt.wdtc.core.download.infterface.ModInstallTaskInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.GameConfig
import org.wdt.wdtc.core.game.GameRuntimeDependency
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.manger.*
import org.wdt.wdtc.core.utils.*
import java.io.File

// TODO Repair and optimization Forge install
class ForgeInstallTask(version: Version, forgeVersion: String) :
	ForgeDownloadInfo(version, forgeVersion), ModInstallTaskInterface {
	
	private val config: GameConfig.Config = version.gameConfig.config
	
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : this(
		version,
		versionsJsonObjectInterface.versionNumber
	)
	
	private fun getCommandLine(index: Int): String {
		val installProfile = forgeInstallProfileJsonObject.processors[index]
		return buildString {
			append(config.javaPath).append(" -cp ")
			installProfile.classpath.forEach {
				it.toGameRuntimeDependency().run {
					append(libraryFilePath).append(";")
				}
			}
			installProfile.libraryName.toGameRuntimeDependency().run {
				libraryFile.getJarManifestInfo("Main-Class").noNull().let {
					append(libraryFilePath).append(STRING_SPACE).append(it).append(" ")
				}
			}
			installProfile.args.forEach { arg ->
				if (arg.startsWith("--")) {
					append(arg).append(STRING_SPACE)
				} else if (arg.startendWith("{", "}")) {
					if (arg == "{BINPATCH}") {
						append(clientLzmaFile.canonicalPath)
					} else {
						dataMap[arg].noNull().let {
							if (it.isRuntime) append(it.client.toGameRuntimeDependency().libraryFilePath)
							else append(it.client)
						}
					}
				}
			}
			
		}
	}
	
	private val dataMap: Map<String, ForgeInstallProfileJsonObject.ForgeDataObject>
		get() {
			return hashMapOf<String, ForgeInstallProfileJsonObject.ForgeDataObject>().apply {
				forgeInstallProfileFile.readFileToJsonObject().getJsonObject("data")
					.asMap()
					.forEach { (key, value) ->
						put("{$key}", value.asJsonObject.parseObject())
					}
				put(
					"{MINECRAFT_JAR}",
					ForgeInstallProfileJsonObject.ForgeDataObject(
						version.versionJar.canonicalPath,
						version.versionJar.canonicalPath
					)
				)
				put("{SIDE}", ForgeInstallProfileJsonObject.ForgeDataObject("client", "server"))
			}
		}
	
	
	private suspend fun downloadClientText() {
		val txtPath = forgeInstallProfileJsonObject.data.let {
			if (it.mojmaps != null) it.mojmaps.client.toGameRuntimeDependency()
			else if (it.mappings != null) it.mappings.client.toGameRuntimeDependency()
			else null
		}.noNull()
		val txtUrl = version.gameVersionJsonObject.downloads.clientMappings.url.let {
			if (isNotOfficialDownloadSource)
				it.toString().replace(pistonDataMojang, currentDownloadSource.dataUrl).toURL().getRedirectUrl().toURL()
			else it
		}
		startDownloadTask(txtUrl, version.gameLibraryDirectory.resolve(txtPath.formJar()))
	}
	
	private suspend fun startInstallForge() {
		forgeInstallProfileJsonObject.processors.let {
			repeat(it.size) { index ->
				it[index].run {
					if (sides != null && sides.first() == "client") {
						startRunCommand(index)
					} else if (args[1] != "DOWNLOAD_MOJMAPS") {
						startRunCommand(index)
					}
				}
			}
		}
	}
	
	private val clientLzmaFile: File = File(wdtcCache, "data/client.lzma")
	
	
	private suspend fun startRunCommand(i: Int) {
		val command = getCommandLine(i)
		logmaker.info("Command Line:$command")
		runOnIO {
			ProcessBuilder("cmd.exe", "/c", command).start().run {
				launch {
					inputStream.reader(charset("GBK")).run {
						while (readlnOrNull().also { logmaker.info(it) } != null);
					}
				}
			}
		}
	}
	
	override fun overwriteVersionJson() {
		version.run {
			gameVersionJsonObject.apply {
				forgeVersionJsonObject.let {
					this.arguments.run {
						it.arguments.jvmList?.forEach { element ->
							jvmList?.add(element)
						}
						it.arguments.gameList?.forEach { element ->
							
							gameList?.add(element)
						}
					}
					this.libraries.addAll(it.libraries)
				}
			}.putToVersionJson()
		}
	}
	
	override fun writeVersionJsonPatches() {
		version.run {
			gameVersionJsonObject.apply {
				patches = mutableListOf(
					versionJson.readFileToJsonObject(),
					forgeVersionJsonFile.readFileToJsonObject()
				)
			}.putToVersionJson()
		}
	}
	
	override suspend fun afterDownloadTask() = coroutineScope {
		launch { startDownloadForgeLibraryFile() }
		launch { downloadClientText() }
		startInstallForge()
	}
	
	override suspend fun beforInstallTask(): Unit = coroutineScope {
		startDownloadInstallJar()
		launch { unZipToInstallProfile() }
		launch { unZipToForgeVersionJson() }
		launch { unZipBySpecifyFile(forgeInstallJarFile, clientLzmaFile) }
	}
	
	private fun String.toGameRuntimeDependency(): GameRuntimeDependency {
		return GameRuntimeDependency(
			this.formatString.let {
				if (isHasFileExtension) cleanStrInString("@$getFileExtension") else it
			}
		).apply {
			if (isHasFileExtension) fileExtension = getFileExtension
			libraryDirectory = this@ForgeInstallTask.version.gameLibraryDirectory
		}
	}
	
	private val String.isHasFileExtension: Boolean
		get() = this.indexOf("@") == -1
	
	private val String.getFileExtension: String
		get() = this.substringAfter("@")
	
	private val String.formatString
		get() = this.cleanStrInString("[").cleanStrInString("]").cleanStrInString("{").cleanStrInString("}")
	
	
	private fun startDownloadForgeLibraryFile() {
		forgeVersionJsonObject.libraries.run {
			val speed = ProgressManger(size).apply {
				coroutineScope = executorCoroutineScope(name = "Download forge library file")
			}
			
			speed.await()
		}
	}
}
